File: 0131-palindrome-partitioning/0131-palindrome-partitioning.ts
function partition(s: string): string[][] {
    const len: number = s.length;
    const result: string[][] = [];

    // dp[i][j] = s[i...j]가 팰린드롬인지 저장
    const dp: boolean[][] = Array.from(
        { length: len },
        () => Array(len).fill(false)
    );

    // 길이가 1인 문자열
    for (let i = 0; i < len; i++) {
        dp[i][i] = true;
    }

    // 길이가 2인 문자열
    for (let i = 0; i < len - 1; i++) {
        dp[i][i + 1] = s[i] === s[i + 1];
    }

    // 길이가 3 이상인 문자열
    for (let length = 3; length <= len; length++) {
        for (let start = 0; start <= len - length; start++) {
            const end: number = start + length - 1;

            if (
                s[start] === s[end] &&
                dp[start + 1][end - 1]
            ) {
                dp[start][end] = true;
            }
        }
    }

    // 현재 분할 결과
    const path: string[] = [];

    // DFS + 백트래킹
    function dfs(start: number): void {
        // 문자열 끝까지 도달
        if (start === len) {
            result.push([...path]);
            return;
        }

        // start부터 가능한 팰린드롬 탐색
        for (let end = start; end < len; end++) {
            if (dp[start][end]) {
                path.push(s.substring(start, end + 1));

                dfs(end + 1);

                // 백트래킹
                path.pop();
            }
        }
    }

    dfs(0);

    return result;
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    static String gs;
    static int len;
    static int dp[][];
    static List<List<String>> result;
    public List<List<String>> partition(String s) {
        gs = s;
        len = gs.length();
        result = new ArrayList<>();
        dp = new int[len][len];
        
        for (int i = 0; i < len; i++) dp[i][i] = 1;
        
        for (int i = 0; i < len - 1; i++)
            if (s.charAt(i) == s.charAt(i + 1)) dp[i][i + 1] = 1;
            else dp[i][i + 1] = 0;
            
        for (int i = 3; i < len + 1; i++) {
            for (int j = 0; j < len - i + 1; j++) {
                int start = j;
                int end = j + i - 1;
                if (s.charAt(start) == s.charAt(end) && dp[start + 1][end - 1] == 1) dp[start][end] = 1;
            }
        }
        
        dfs("",0);
        return result;
    }
    static void dfs(String substrings, int now){
        for(int i = now;i<len;i++){
            if(dp[now][i] == 1){
                String new_substrings=substrings+gs.substring(now,i+1)+" ";
                if(i == len-1) result.add(Arrays.asList(new_substrings.split(" ")));
                else dfs(new_substrings, i+1);
            }
        }
    }
}